src/main/java/com/iota/iri/network/NeighborRouter.java

```diff
@@ -0,0 +1,866 @@
+        if (!Arrays.equals(byteEncodedCooAddress, handshake.getByteEncodedCooAddress())) {
+            log.error("dropping handshaked connection to neighbor {} as it uses a different coordinator address",
+                    identity);
+            closeNeighborConnection(channel, null, selector);
+            return false;
+        }
+
+        // check whether we support the supported protocol versions by the neighbor
+        int supportedVersion = handshake.isNeighborSupported(Protocol.SUPPORTED_PROTOCOL_VERSIONS);
+        if (supportedVersion <= 0) {
+            log.error(
+                    "dropping handshaked connection to neighbor {} as its highest supported protocol version {} is not supported",
+                    identity, Math.abs(supportedVersion));
+            closeNeighborConnection(channel, null, selector);
+            return false;
+        }
+        neighbor.setProtocolVersion(supportedVersion);
+
+        // after a successful handshake, the neighbor's server socket port is initialized
+        // and thereby the identity of the neighbor is now fully distinguishable
+
+        // check whether the remote server socket port from the origin URI
+        // actually corresponds to the port advertised in the handshake packet
+        int originPort = neighbor.getRemoteServerSocketPort();
+        int handshakePort = handshake.getServerSocketPort();
+        if (originPort != Neighbor.UNKNOWN_REMOTE_SERVER_SOCKET_PORT && originPort != handshakePort) {
+            log.warn("dropping handshaked connection from {} as neighbor advertised "
+                    + "wrong server socket port (wanted {}, got {})", identity, originPort, handshakePort);
+            closeNeighborConnection(channel, null, selector);
+            return false;
+        }
+        neighbor.setRemoteServerSocketPort(handshakePort);
+
+        // check if neighbor is already connected
+        String newIdentity = neighbor.getHostAddressAndPort();
+        if (connectedNeighbors.containsKey(newIdentity)) {
+            log.info("dropping handshaked connection from {} as neighbor is already connected", newIdentity);
+            // pass just host address to not actually delete the already existing connection/neighbor
+            closeNeighborConnection(channel, null, selector);
+            return false;
+        }
+
+        // check if the given host + server socket port combination is actually defined in the config/wanted
+        if (!config.isAutoTetheringEnabled() && !allowedNeighbors.contains(newIdentity)) {
+            log.info("dropping handshaked connection as neighbor from {} is not allowed to connect", newIdentity);
+            closeNeighborConnection(channel, null, selector);
+            return false;
+        }
+
+        log.info("neighbor connection to {} is ready for messages [latency {} ms, protocol version {}]", newIdentity,
+                System.currentTimeMillis() - handshake.getSentTimestamp(), supportedVersion);
```

Comment:
Won't remove because it can still add value if there's actually a long distance between two nodes. Neighbors are not just added and removed in the beginning of a node startup.